File: src/TypeFormatter/PrimitiveUnionTypeFormatter.ts
import { TypeTJSGError } from "../Error/Errors.js";
import type { Definition } from "../Schema/Definition.js";
import type { RawTypeName } from "../Schema/RawType.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import type { BaseType } from "../Type/BaseType.js";
import { BooleanType } from "../Type/BooleanType.js";
import { NullType } from "../Type/NullType.js";
import { NumberType } from "../Type/NumberType.js";
import { PrimitiveType } from "../Type/PrimitiveType.js";
import { StringType } from "../Type/StringType.js";
import { UnionType } from "../Type/UnionType.js";
import { uniqueArray } from "../Utils/uniqueArray.js";

export class PrimitiveUnionTypeFormatter implements SubTypeFormatter {
    public supportsType(type: BaseType): boolean {
        return type instanceof UnionType && type.getTypes().length > 0 && this.isPrimitiveUnion(type);
    }
    public getDefinition(type: UnionType): Definition {
        return {
            type: uniqueArray(type.getTypes().map((item) => this.getPrimitiveType(item))),
        };
    }
    public getChildren(type: UnionType): BaseType[] {
        return [];
    }

    protected isPrimitiveUnion(type: UnionType): boolean {
        return type.getTypes().every((item) => item instanceof PrimitiveType);
    }

    protected getPrimitiveType(item: BaseType): RawTypeName {
        if (item instanceof StringType) {
            return "string";
        }

        if (item instanceof NumberType) {
            return "number";
        }

        if (item instanceof BooleanType) {
            return "boolean";
        }

        if (item instanceof NullType) {
            return "null";
        }

        throw new TypeTJSGError("Unexpected code branch", item);
    }
}

File: internal/formatter/primitive_union_type_formatter.go
package formatter

import (
	"fmt"

	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// PrimitiveUnionTypeFormatter mirrors src/TypeFormatter/PrimitiveUnionTypeFormatter.ts.
type PrimitiveUnionTypeFormatter struct{}

func NewPrimitiveUnionTypeFormatter() *PrimitiveUnionTypeFormatter {
	return &PrimitiveUnionTypeFormatter{}
}

func (f *PrimitiveUnionTypeFormatter) SupportsType(t types.Type) bool {
	unionType, ok := t.(*types.UnionType)
	return ok && len(unionType.Types()) > 0 && f.isPrimitiveUnion(unionType)
}

func (f *PrimitiveUnionTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	unionType := t.(*types.UnionType)
	names := make([]string, 0, len(unionType.Types()))
	for _, item := range unionType.Types() {
		names = append(names, f.getPrimitiveType(item))
	}
	return &schema.Definition{Type: uniqueStrings(names)}
}

func (f *PrimitiveUnionTypeFormatter) GetChildren(t types.Type) []types.Type { return nil }

func (f *PrimitiveUnionTypeFormatter) isPrimitiveUnion(unionType *types.UnionType) bool {
	for _, item := range unionType.Types() {
		switch item.(type) {
		case *types.StringType, *types.NumberType, *types.BooleanType, *types.NullType:
		default:
			return false
		}
	}
	return true
}

func (f *PrimitiveUnionTypeFormatter) getPrimitiveType(item types.Type) string {
	switch item.(type) {
	case *types.StringType:
		return "string"
	case *types.NumberType:
		return "number"
	case *types.BooleanType:
		return "boolean"
	case *types.NullType:
		return "null"
	}
	panic(fmt.Errorf("unexpected code branch for type %T", item))
}
